File: DeviseAlgorithm/test_2/Fibonacci/Dynamic/FibnacciDP.c
/*
    斐波那契数动态规划法(Dynamic Programming)
	学号：20201050470
	时间复杂度：O(n)

*/
#include<stdio.h>
#include<stdlib.h>
/*动态规划法*/ 
int *DP_f(int n){
	int *f=(int*)malloc(sizeof(int)*n);
	f[0]=0;
	for(int i=1;i<=n;i++){
		if(i==1||i==2){
			f[i]=1;
		}
		else{
			f[i]=f[i-1]+f[i-2];
		}
	}
	return f;//返回fibonacci数列
}

int main(){
	int n;
	printf("输入测试数据: ");
	scanf("%d",&n);
	printf("输出结果为：");
	int *f=DP_f(n);
	for(int i=1;i<=n;i++){
		printf("%d ",f[i]);
	}
	return 0;
}

File: DeviseAlgorithm/test_2/Fibonacci/Dynamic/FibnacciDP.java
/*  
     Java斐波那契数列动态规划(Dynamic Programming)
     时间复杂度：O(n)
*/
import java.util.Scanner;

public class FibnacciDP {
    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("请输入斐波那契数列的项数：");
            int n = sc.nextInt();
            DPF dp=new DPF();
            System.out.print("动态规划结果为：");
            for(int i=1;i<=n;i++){
                System.out.print(dp.fib(i)+" ");
            }
        }
    }
}

class DPF{
    public int fib(int n){
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }
}